lib/ansible/modules/network/avi/avi_api_session.py

```diff
@@ -0,0 +1,187 @@
+        data_json=dict(type='str'),
+        timeout=dict(default=60)
+    )
+    argument_specs.update(avi_common_argument_spec())
+    module = AnsibleModule(argument_spec=argument_specs)
+
+    if not HAS_AVI:
+        return module.fail_json(msg=(
+            'Avi API SDK is not installed. '
+            'Use "sudo pip install --upgrade avisdk" to install Avi SDK. '
+            'For more details visit https://github.com/avinetworks/sdk.'))
+    tenant_uuid = module.params.get('tenant_uuid', None)
+    api = ApiSession.get_session(
+        module.params['controller'], module.params['username'],
+        module.params['password'], tenant=module.params['tenant'],
+        tenant_uuid=tenant_uuid)
+
+    tenant = module.params.get('tenant', '')
+    timeout = int(module.params.get('timeout', 60))
+    path = module.params.get('path', '')
+    params = module.params.get('params', None)
+    data = module.params.get('data', None)
+    if ((data is None) and
+            (module.params.get('data_json', None) is not None)):
+        data = json.loads(module.params.get('data_json', None))
+    method = module.params['http_method']
+
+    existing_obj = None
+    changed = method != 'get'
+    if method == 'put':
+        gparams = deepcopy(params) if params else {}
+        gparams.update({'include_refs': '', 'include_name': ''})
+        rsp = api.get(path, tenant=tenant, tenant_uuid=tenant_uuid,
+                      params=gparams)
+        if rsp.status_code == 404:
+            method = 'post'
+        else:
+            existing_obj = rsp.json()
+            changed = not avi_obj_cmp(data, existing_obj)
+            cleanup_absent_fields(data)
+    if method == 'patch':
+        gparams = deepcopy(params) if params else {}
+        gparams.update({'include_refs': '', 'include_name': ''})
+        rsp = api.get(path, tenant=tenant, tenant_uuid=tenant_uuid,
+                      params=gparams)
+        existing_obj = rsp.json()
+    fn = getattr(api, method)
+    rsp = fn(path, tenant=tenant, tenant_uuid=tenant, timeout=timeout,
+             params=params, data=data)
+    if method == 'delete' and rsp.status_code == 404:
+        changed = False
+        rsp.status_code = 200
+    if method == 'patch' and existing_obj and rsp.status_code < 299:
+        # Ideally the comparison should happen with the return values
+        # from the patch API call. However, currently Avi API are
+        # returning different hostname when GET is used vs Patch.
+        # tracked as AV-12561
+        if path.startswith('pool'):
+            time.sleep(1)
+        gparams = deepcopy(params) if params else {}
+        gparams.update({'include_refs': '', 'include_name': ''})
+        rsp = api.get(path, tenant=tenant, tenant_uuid=tenant_uuid,
+                      params=gparams)
+        new_obj = rsp.json()
+        changed = not avi_obj_cmp(new_obj, existing_obj)
+    return ansible_return(module, rsp, changed, req=data)
+
+
```

Comment:
Actually, this is the only option when you want to use HTTP patch semantics where you don't have the full object. This is actively used. I can provide references to actual field use cases if it would help.